cli/azd/pkg/azsdk/zip_deploy_client.go

```diff
@@ -113,8 +117,87 @@ func (c *ZipDeployClient) BeginDeploy(
 	return runtime.NewPoller(response, c.pipeline, pollerOptions)
 }
 
+// Deploys the specified application zip to the azure app service using deployment status api and waits for completion
+func (c *ZipDeployClient) BeginDeployTrackStatus(
+	ctx context.Context,
+	zipFile io.Reader,
+	subscriptionId,
+	resourceGroup,
+	appName string,
+	printStatus func(string),
+) (*runtime.Poller[armappservice.WebAppsClientGetProductionSiteDeploymentStatusResponse], error) {
+	printStatus("Creating deployment request")
+	request, err := c.createDeployRequest(ctx, zipFile)
+	if err != nil {
+		return nil, err
+	}
+
+	response, err := c.pipeline.Do(request)
+	if err != nil {
+		return nil, err
+	}
+
+	defer response.Body.Close()
+
+	if !runtime.HasStatusCode(response, http.StatusAccepted) {
+		return nil, runtime.NewResponseError(response)
+	}
+
+	client, err := armappservice.NewWebAppsClient(subscriptionId, c.cred, nil)
+	if err != nil {
+		return nil, fmt.Errorf("creating web app client: %w", err)
+	}
+
+	printStatus("Getting deployment status id")
+	deploymentStatusId := response.Header.Get("Scm-Deployment-Id")
+	matched, err := regexp.MatchString(`^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$`, deploymentStatusId)
+	if err != nil {
+		return nil, err
+	}
+
+	if !matched {
+		return nil, fmt.Errorf("invalid or empty deployment status id")
+	}
+
+	// nolint:lll
+	// Example definition: https://github.com/Azure/azure-rest-api-specs/tree/main/specification/web/resource-manager/Microsoft.Web/stable/2022-03-01/examples/GetSiteDeploymentStatus.json
+	printStatus("Getting deployment status response")
+	poller, err := client.BeginGetProductionSiteDeploymentStatus(ctx, resourceGroup, appName, deploymentStatusId, nil)
+	if err != nil {
+		return nil, fmt.Errorf("getting deployment status: %w", err)
+	}
+
+	return poller, nil
+}
+
+func (c *ZipDeployClient) DeployTrackStatus(
+	ctx context.Context,
+	zipFile io.Reader,
+	subscriptionId string,
+	resourceGroup string,
+	appName string,
+	printStatus func(string)) (armappservice.WebAppsClientGetProductionSiteDeploymentStatusResponse, error) {
+	var response armappservice.WebAppsClientGetProductionSiteDeploymentStatusResponse
+
+	poller, err := c.BeginDeployTrackStatus(ctx, zipFile, subscriptionId, resourceGroup, appName, printStatus)
+	if err != nil {
+		return response, err
+	}
+
+	response, err = poller.PollUntilDone(ctx, &runtime.PollUntilDoneOptions{
```

Comment:
In a future iteration (after we land this), I suspect this is what will need to change such that we can call `printStatus` on each turn of the crank of the poller if the status has changed.

It's sort of sad that there isn't some sort of way to register an observer that would let you observe the response of each `Poll` call.  If we had that, it would be easy for us to plug in the status observer here and continue to leverage all the PollUntilDone logic around delays instead of us having to open-code that so we can get access to the individual status lines during the deployment.